### PR TITLE
PUBSTWO-1722 - SIPP Repairs

### PR DESCRIPTION
**Pull request by dsteinich**
Several Major Issues:
* Transaction boundaries incorrect
* XML serializing nil to empty string
* ORCID validation of check digit
* ORCID should not prevent addition of publication from SIPP
* Errors & messages not propagating to correct levels or stopping processing when appropriate
* Missing parent validations allowing attempts to persist bad data

### CODE REVIEW

File: src/main/java/gov/usgs/cida/pubs/busservice/ext/ExtPublicationService.java [hunk 1]
```diff
@@ -28,22 +30,32 @@ public ExtPublicationService (
...
+		MpPublication rtn;
+		ValidationResults validationErrors = new ValidationResults();
+		validationErrors.addValidationResults(extPublicationContributorBusService.processPublicationContributors(mpPublication.getContributors()));
+		validationErrors.addValidationResults(processCostCenters(mpPublication.getCostCenters()));
+		if (validationErrors.isValid()) {
+			rtn = pubBusService.createObject(mpPublication);
+		} else {
+			rtn = mpPublication;
+		}
+		mpPublication.addValidationResults(validationErrors);
```
Comment (orosz-usgs) [inline src/main/java/gov/usgs/cida/pubs/busservice/ext/ExtPublicationService.java]: Calling addValidationResults() on the passed in parameter is confusing. Probably should be called on
rtn

File: src/main/java/gov/usgs/cida/pubs/validation/mp/parent/ParentExistsValidatorForMpPublicationContributor.java [hunk 2]
```diff
@@ -27,30 +27,37 @@ public boolean isValid(PublicationContributor<?> value, ConstraintValidatorConte
...
 					.addPropertyNode("contributor").addConstraintViolation();
 			}
 
-			if (null != value.getContributorType() && null != value.getContributorType().getId() 
-					&& null == ContributorType.getDao().getById(value.getContributorType().getId())) {
+			if (null == value.getContributorType()
+					|| (null != value.getContributorType()
+						&& (null == value.getContributorType().getId()
+							|| (null != value.getContributorType().getId()
+								&& null == ContributorType.getDao().getById(value.getContributorType().getId()))))) {
```
Comment (orosz-usgs) [inline src/main/java/gov/usgs/cida/pubs/validation/mp/parent/ParentExistsValidatorForMpPublicationContributor.java]: This if seems a little hard to follow . It could be broken into 
if (null == value.getContributorType() || null == value.getContributorType().getId()
   || null == ContributorType.getDao().getById(value.getContributorType().getId())) {
     rtn = false
line 4=37 as well

File: src/main/java/gov/usgs/cida/pubs/validation/mp/parent/ParentExistsValidatorForMpPublicationCostCenter.java [hunk 3]
```diff
@@ -33,15 +33,18 @@ public boolean isValid(PublicationCostCenter<?> value, ConstraintValidatorContex
 					.addPropertyNode("publicationId").addConstraintViolation();
 			}
 			
-			if (null != value.getCostCenter() && null != value.getCostCenter().getId() 
-					&& null == CostCenter.getDao().getById(value.getCostCenter().getId())) {
+			if (null == value.getCostCenter()
+					|| (null != value.getCostCenter()
+						&& (null == value.getCostCenter().getId()
+							|| (null != value.getCostCenter().getId()
+								&& null == CostCenter.getDao().getById(value.getCostCenter().getId()))))) {
```
Comment (orosz-usgs) [inline src/main/java/gov/usgs/cida/pubs/validation/mp/parent/ParentExistsValidatorForMpPublicationCostCenter.java]: similar to previous comment

File: src/main/java/gov/usgs/cida/pubs/validation/mp/parent/ParentExistsValidatorForMpPublicationLink.java [hunk 4]
```diff
@@ -35,8 +35,11 @@ public boolean isValid(PublicationLink<?> value, ConstraintValidatorContext cont
 					.addPropertyNode("publicationId").addConstraintViolation();
 			}
 	
-			if (null != value.getLinkType() && null != value.getLinkType().getId()
-					&& null == LinkType.getDao().getById(value.getLinkType().getId())) {
+			if (null == value.getLinkType()
+					|| (null != value.getLinkType()
+						&& (null == value.getLinkType().getId()
+							|| (null != value.getLinkType().getId()
+								&& null == LinkType.getDao().getById(value.getLinkType().getId()))))) {
```
Comment (orosz-usgs) [inline src/main/java/gov/usgs/cida/pubs/validation/mp/parent/ParentExistsValidatorForMpPublicationLink.java]: similar to previous comment

**Review by orosz-usgs — APPROVED** [review]

**Review by mbucknell — APPROVED** [review]
